Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -1171,7 +1171,7 @@ def try_to_parse_directives(self, node):
 
     def try_to_parse_directive(self, optname, args, kwds, pos):
         if optname == 'np_pythran' and not self.context.cpp:
-            raise PostParseError(pos, 'The %s directive can only be used in C++ mode.' % optname)
+            return optname, PostParseError(pos, 'The %s directive can only be used in C++ mode.' % optname)
```

Comment:
This PR needs to avoid raising an `PostParseError` only in case of `with gil()` or `with nogil()`. All decorators (including `@gil`/`@nogil`) should raise an PostParseError as it was before. This was rather difficult to decide inside `try_to_parse_directive()` since we need an information that the directive is inside with statement. Hence, I decided to circumvent it by simply returning error as value and deciding whether should be raised in a place where I have information that it is inside `with` statement.

I think this should be commented in the code, but I would like to see tests passing and some early feedback whether it makes sense or not.

Reply:
I'm a little sceptical of this approach. It just seems very non-obvious to me.

My (untested) suggested alternatives
1) I wonder if it'd be better to pick out `nogil`/`gil` `CallNodes` earlier (at the top of `visit_WithStatNode`), and handle them there without putting them through `try_to_parse_directives`.
2) Add a `in_withstate` boolean argument to `try_to_parse_directives`, and use that to skip most of the work for the `nogil`/`gil` `CallNodes` there (i.e skip `try_to_parse_directive` in that case).
3) Have an optional `catch` in `try_to_parse_directives` that catches `PostParseException` only if the name is `nogil`/`gil` (this is a variant of option "2" but does at least run `try_to_parse_directives` if possible. I'm not sure if there's value in that).

I haven't actually tried these options so there may be something I'm missing. But that's probably what I'd try.

Reply:
I agree that returning an exception (in addition to a normal value) is counter-intuitive. If something is an error, it's an error and not just half an error that we may forget to raise later.

I think we should try to keep the property that `try_to_parse_directive()` simply returns a name-value pair of some sort (or raises an exception). It might be a simple value or an expression, but some kind of value. We can then still reject (non-constant) expressions where they are not allowed.

Reply:
Basically, if `try_to_parse_directive()` cannot decide whether arguments are allowed or not, and whether they need to be constant or not, then the caller should take the decision, but afterwards. There is no need to complicate the decisions taken _inside_ of the method just because the method is used in different contexts.

Reply:
I agree and I have added new parameter for optnames allowed to have expressions. Let me know if now it is better.